bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java

```diff
@@ -1331,31 +1049,200 @@ private void updateChildThingStatusForDisabledBridges(Bridge bridge) {
     private void persistThingEnableStatus(ThingUID thingUID, boolean enabled) {
         logger.debug("Thing with UID {} will be persisted as {}.", thingUID, enabled ? "enabled." : "disabled.");
         if (enabled) {
-            // Clear the disabled thing storage. Otherwise the handler will NOT be initialized later.
-            storage.remove(thingUID.getAsString());
+            // Clear the disabled thing storage. Otherwise, the handler will NOT be initialized later.
+            disabledStorage.remove(thingUID.getAsString());
         } else {
             // Mark the thing as disabled in the storage.
-            storage.put(thingUID.getAsString(), "");
+            disabledStorage.put(thingUID.getAsString(), "");
         }
     }
 
     @Override
     public boolean isEnabled(ThingUID thingUID) {
-        Thing thing = this.things.get(thingUID);
+        Thing thing = things.get(thingUID);
         if (thing != null) {
             return thing.isEnabled();
         }
 
         logger.debug("Thing with UID {} is unknown. Will try to get the enabled status from the persistent storage.",
                 thingUID);
-        return !isDisabledByStorage(thingUID);
+        return !disabledStorage.containsKey(thingUID.getAsString());
     }
 
-    private boolean isDisabledByStorage(ThingUID thingUID) {
-        return storage.containsKey(thingUID.getAsString());
+    private boolean checkAndPerformUpdate(Thing thing, ThingHandlerFactory factory) {
+        final int currentThingTypeVersion = Integer
+                .parseInt(thing.getProperties().getOrDefault(PROPERTY_THING_TYPE_VERSION, "0"));
+
+        UpdateInstructionKey thingKey = new UpdateInstructionKey(factory, thing.getThingTypeUID().getId());
+        List<ThingUpdateInstruction> instructions = updateInstructions.getOrDefault(thingKey, List.of()).stream()
```

Comment:
How about sorting them by version?

Reply:
I don't think that's a good idea. Especially if we use REMOVE/ADD for updating channels without configuration, we have to preserve the original order (at least for these instructions), otherwise the builder will fail. So it's safer to not mess with the order.

Reply:
Ok, let me precise my question:
Shouldn't they sorted by `targetVersion`, so that we can be sure that changes for version 2 are never applied before the changes for version 1? People might add new information in the XSD always to the top, we cannot guarantee any order here.
I'm fine with keeping the order of instructions for the same target version, though.

Reply:
Done.